### DISCUSSION TITLE
/hyndsight/transformations/

### DISCUSSION THREAD

**Discussion by robjhyndman** [robjhyndman/robjhyndman.com#81] [Announcements]
Transforming data with zeros 
 https://robjhyndman.com/hyndsight/transformations/

**Comment by robjhyndman** [comment]
```header
Originally posted by John Ramey on 2011-06-10 19:41:13
````
Have you considered the Yeo-Johnson power transformation? It was
published in Biometrika in 2000, and the R package VGAM includes a
yeo.johnson function. Also, it is an extension of Box Cox.

It was designed for the special cases that you have mentioned.

Do you have any suggestions on multivariate approaches to the same issue
without resorting to a series of univariate transformations?

Thanks.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2011-06-11 03:17:58
````
I didn\'t know about that family of transformations. Thanks for the tip.

I\'ve never considered the problem for multivariate data.

**Reply by robjhyndman** [reply]
```header
Originally posted by Matt Dancho on 2018-02-18 16:20:25
````
I agree. Yeo-Johnson solves most of the issues from non-negative data
and zeros that are problematic for other transformations such as log,
sqrt, and Box Cox.

**Comment by robjhyndman** [comment]
```header
Originally posted by isomorphisms on 2012-07-17 07:30:51
````
What about started logs?

**Comment by robjhyndman** [comment]
```header
Originally posted by Ellie_K on 2012-07-18 22:22:15
````
Might you post the specific URL for the inquiry on Stats StackExchange
(I guess I should say &quot;Cross Validated&quot;) that you mentioned? I use
that site, and would like to check if there have been any additional
responses in the 2 year interim. Thank you for considering my request!

**Reply by robjhyndman** [reply]
```header
Originally posted on 2012-07-18 22:32:16
````
[http://stats.stackexchange\....](http://stats.stackexchange.com/q/1444/159)

**Reply by robjhyndman** [reply]
```header
Originally posted by Ellie_K on 2012-08-03 04:47:05
````
Thank you, Rob!

**Comment by robjhyndman** [comment]
```header
Originally posted by A.F.M. Kamal Chowdhury on 2013-09-02 04:31:09
````
Hi Rob,

Thank you for this very useful post. I have been working with a 60 years
daily rainfall data which contains many zero values. I have tried
&quot;[boxcox.fit](http://boxcox.fit &quot;boxcox.fit&quot;)()
func­tion of pack­age geoR&quot; as you mentioned above to transform my data,
but have obtained an error massage like: &quot;Error: could not find
function
&quot;[boxcox.fit](http://boxcox.fit &quot;boxcox.fit&quot;)&quot;.
Can you please help me in this regard?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-09-02 05:41:07
````
It sounds like you haven\'t loaded the package. Use library(geoR)

**Reply by robjhyndman** [reply]
```header
Originally posted by AFM Kamal Chowdhury on 2013-09-02 12:22:47
````
Thank you for your reply. It seems
[boxcox.fit](http://boxcox.fit &quot;boxcox.fit&quot;)()
function has been changed into boxcoxfit(). Even trying with the later
one, I have got error massage as below:

&quot;\&gt; boxcoxfit(NARCliM_rainfall_ts)

Error in boxcoxfit(NARCliM_rainfall_ts) :\
Transformation requires positive data&quot;

As I have mentioned earlier, I have zero values in my data.

AFM Kamal Chowdhury\
MPhil Student\
University of Newcastle, Australia

**Comment by robjhyndman** [comment]
```header
Originally posted by Elaine Oon on 2013-10-10 20:00:27
````
Hi Rob

I found your post on box-cox transformation above very helpful. I have a
query which I would like to seek your advice: I am trying to do a
box-cox transformation with swift. I have a dependent variable, annual
foreign sales of companies (in US\$ thousands) which contains zeros, for
a set of panel data. I have been advised to add a small amount, for
example, 0.00001 to the annual foreign sales figures so that I can take
the log, but I think box-cox transformation will produce a more
appropriate constant than 0.00001. I have done a box-cox transformation
on R with the codes below, but it has given me a very large lambda2 of
31162.8.

library(geoR)\
boxcoxfit(bornp\$ForeignSales, lambda2 = TRUE)\
#R output - Fitted parameters:\
\# lambda lambda2 beta sigmasq\
\# -1.023463e+00 3.116280e+04 9.770577e-01 7.140328e-11

My hunch is that the above value of lambda2 is very large, so I am not
sure if I need to run the boxcoxfit with my independent variables like
below:\
boxcoxfit(bornp\$ForeignSales, bornp\$family bornp\$roa
bornp\$solvencyratio,\
lambda2=TRUE)

I am still trying to identify the best set of independent variables, so
I am not sure if using the boxcoxfit with independent variables at this
stage will work or is best.

I would be very grateful for your advice on the above.

Thanks,\
Elaine

**Comment by robjhyndman** [comment]
```header
Originally posted by Carlos Eduardo on 2014-04-04 15:41:41
````
Hi. My model has two independent variables. One of them has zero as a
valid value. Is it sound practice to split the dataset in two? The first
one with data with both independent variables present non-zero values.
The second one with data where the independent variable has zero value.
Then, perform a regression on each dataset so the resulting model is
comprised of two different equations. Is there some measure I should be
aware of about using this appoach? Does it have a name? Thanks in
advance.

**Comment by robjhyndman** [comment]
```header
Originally posted by Don Strong on 2014-04-28 19:53:58
````
David I. Warton and Francis K. C. Hui 2011. The arcsine is asinine: the
analysis of proportions in ecology. Ecology
92:3--10.http://[dx.doi.org/10.1890/10-0340.1](http://dx.doi.org/10.1890/10-0340.1 &quot;dx.doi.org/10.1890/10-0340.1&quot;)

**Comment by robjhyndman** [comment]
```header
Originally posted by Daniel on 2014-04-28 20:20:47
````
Hi Rob,

Very useful post, thanks. Do you know if there is any discussion of the
practical consequences of the various choices e.g. in terms of
bias/variance?

A small comment: the last header should probably read &quot;mixture
models&quot;. &quot;Mixed models&quot; means random effects to me rather than zero
inflation.

Best, Daniel\

**Comment by robjhyndman** [comment]
```header
Originally posted by K Denny on 2014-04-28 20:32:26
````
Finding a transformation that permits zeros seems to me to be a
&quot;mindless&quot; way of handling the issue. Mixture models at least try to
model why there might be a bunch of zeros. In econometrics where you are
often dealing with zeros, like income, the norm is to use a censored
normal/Tobit or a Heckman selection model or one of their
generalisations. For count data, a hurdle model makes sense or a
zero-inflated model.

**Comment by robjhyndman** [comment]
```header
Originally posted by Guest on 2015-01-09 05:10:10
````
Thanks for this post and for your related comments on Cross Validated
Rob! I gave the inverse sinh transformation a try on some continuous
zero inflated data that I have and it produced a nice distribution
except for the zero inflation. On the other hand, a log(y+alpha)
transform is a little rougher, but solves the zero inflation fairly
nicely.

**Comment by robjhyndman** [comment]
```header
Originally posted by AngeloDAmbrosio on 2015-10-29 18:37:40
````
Hi! William Gould, president of StataCorpm (and some others), says that
when you have a non negative dependent variable is ok to use Poisson
regression even if the generation process is not Poisson and the 0s are
not generated from a different process (like in the mixture cases you
shown). The estimates are correct anyway. Just the standard errors are
too small but you can use a more robust variance estimator or, like I
do, use bootstrap for the error and avoid the problem altogether. Also
negative binomial and quasi poisson are good alternatives for the
estimation of errors.\
[http://blog.stata.com/2011/\...](http://blog.stata.com/2011/08/22/use-poisson-rather-than-regress-tell-a-friend/ &quot;http://blog.stata.com/2011/08/22/use-poisson-rather-than-regress-tell-a-friend/&quot;)

**Comment by robjhyndman** [comment]
```header
Originally posted by online essay writers on 2016-04-04 13:50:02
````
This technique that you have shared looks pretty helpful and really
useful for them to have that will give them the knowledge and another
form of technique that will going to help those people who might need to
have this kind of idea for their output or work.

**Comment by robjhyndman** [comment]
```header
Originally posted by Lino Llamosa Mejia on 2016-05-13 15:37:32
````
Dear Rob J. Hyndman

Thank you for your contribution on this field. What happen when you have
the following equation on log: ln(Y) = Ln(T) + Ln(M) + Ln(K), on which Y
represent Production, T Number of worker, M intermediary material and K
capital for a dataset of 3630 observations. The problem is that I have
400 zero values observation of variable=Number of worker. Can I treat
with Box-Cox transformation only 1 of those 3 variable? Mean, treat only
(T) with Box-Cox and the rest continuous with logarithm natural
transformation?

Thank for you value comment

**Comment by robjhyndman** [comment]
```header
Originally posted by Dan Lewer on 2017-03-10 13:00:05
````
Thanks so much for this. It\'s really helpful. I\'m sorry to ask a basic
question - but do you know how you interpret the intercept and
coefficients if you use the log(y + 1) for the independent variable in
linear regression? I\'ve been searching around the internet and am
struggling a bit\...

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-03-12 21:09:28
````
No. If y is large, the interpretation is approximately as for logs. But
otherwise, I don\'t think the parameters have a neat interpretation.

**Comment by robjhyndman** [comment]
```header
Originally posted by gkampolis on 2018-08-15 15:45:04
````
Hi Dr. Hyndman,

Quick question: in the simple case of log(y+1), when we fit and forecast
with a model, is it enough to simply subtract the 1 in the results
(mean, lower &amp; upper) to shift them back down or am I forgetting
something?

For example:\
fit \&lt;- ets(y + 1, lambda = 0, biasadj = TRUE)\
for \&lt;- forecast(fit, h = horizon, biasadj = TRUE)

and then

for\$mean \&lt;- for\$mean -1\
for\$upper \&lt;- for\$upper -1\
for\$lower \&lt;- for\$lower -1

Many thanks in advance for your time.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2018-08-15 22:30:12
````
Yes, that should work.

**Reply by robjhyndman** [reply]
```header
Originally posted by gkampolis on 2018-08-16 11:50:29
````
Thank you for getting back to me and for your time.\
Kind regards,\
George

**Comment by alberto-agudo** [comment]
Hi Rob, thank you so much for the detailed analysis above, it's being really helpful! 

I have a question related to the type of data we are dealing with. You mention the advantage of using mixture models for continuous data with big spikes at zero. However, you include insurance claims in continuous data. I have a similar variable to insurance claims (i.e., an integer), which is number of ads to which a user has been exposed in a certain week. 

This usually distributes as a variable whose third quartile for some age-gender segments is still 0 (the median is zero for all my segments) for all the weeks in the dataset, so I could see the advantage of using mixture models because of the spike at zero. However, the variable gets at most at around 5 exposures per week (with some outliers with around 50 ad exposures). In this case of a continuous but integer variable, would mixture models still be a good option to deal with skewness? 

Thank you very much!

Best wishes,
Alberto.

**Reply by robjhyndman** [reply]
Yes, that can be handled with a mixture distribution, but with the component distribution being on counts rather than continuous. This type of mixture is often called "zero-inflated". That is, it is a count distribution, with more zeros than you might expect from a typical count model. So you can have zero-inflated Poisson models, for example.

**Comment by AngelPone** [comment]
Hi Rob, thank for your valuable posts.
When forecasting using auto.arima and ets with BoxCox transformation, it is possible that, sometimes, the fitted value and forecast value on the transformed scale are slightly bigger than observations, but the fitted mean and forecast mean on the original scale (with bias adjustments) are unreasonably larger than observations.

Here is an example from the tourism dataset:
![](https://ponezimg.oss-cn-beijing.aliyuncs.com/uPic/e7skGO.png)
How do you handle this case? In other word, how do you determine if use lambda or not (without out-of-sample evaluation)?

**Reply by robjhyndman** [reply]
That happens when the transformation is too strong. I won't use an automated procedure for choosing lambda because this sort of thing can happen. If I use a transformation, I'll err on making it slightly larger than what might appear optimal.
